apps/examples/board_specific_demo/ameba_secure_storage_demo/secure_storage.c

```diff
@@ -63,88 +69,150 @@ int sstorage_main(int argc, char *argv[])
 	security_handle hnd;
 	security_error res;
 	security_data data;
-	uint32_t slot;
+	uint32_t input_data = DEFAULT_TEST_DATA;
+	uint32_t slot_index = TEST_SLOT_NUM;
 	uint32_t address;
 	uint32_t length = TEST_DATA_LENGTH;
 	char ss_path[7];
+	int option;
+	int i;
+
+	while ((option = getopt(argc, argv, "d:s:h")) > 0) {
+		switch (option) {
+		case 'd':
+			sscanf(optarg, "%x", &input_data);
+			break;
+		case 's':
+			slot_index = atoi(optarg);
+			if (slot_index > 31) {
+				printf("Slot index cannot be greater than 31, normalizing to 0\n");
+				slot_index = 0;
+			}
+			break;
+		default:
+			SHOW_USAGE
+			return 0;
+		}
+	}
 
 	printf("Start testing SE Secure Storage!!\n");
+
+	printf("Data: %2x, slot: %d\n", input_data, slot_index);
 	printf("Start Security init!!\n");
-	sleep(1);
 	res = security_init(&hnd);
 	if (res != SECURITY_OK) {
-		printf("security_init() Fail\n", __LINE__);
+		printf("security_init() Fail\n, %d", __LINE__);
 		sleep(1);
 	}
-
+	printf("Security init complete\n");
 	sleep(1);
-	data.data = (void *)a;
-	data.length = length;
 
-	printf("Start Secure Storage write!!\n");
-	for (slot = 0; slot < TEST_SLOT_NUM; slot++) {
-		snprintf(ss_path, 7, "ss/%d", slot);
-		memset(a, (0xA0 + slot), length);
-		a[length] = '\0';
-		printf("ss_write_secure_storage() Slot: %d\n", slot);
-		sleep(1);
-		res = ss_write_secure_storage(hnd, ss_path, 0, &data);
-		if (res != SECURITY_OK) {
-			printf("ss_write_secure_storage() Fail\n", __LINE__);
+	data.length = TEST_DATA_LENGTH;
+
+	/* Form the SS path name based on the slot index where we want to read and write */
+	snprintf(ss_path, 7, "ss/%d", slot_index);
+
+	printf("Let us read once before writing to the slot\n");
+	printf("ss_read_secure_storage() Slot: %d\n\n", slot_index);
+	res = ss_read_secure_storage(hnd, ss_path, 0, &data);
+	if (res != SECURITY_OK) {
+		printf("ss_read_secure_storage() Fail, %d, res = %d\n", __LINE__, res);
+	} else {
+		printf("data.length: %d\n", data.length);
+		memcpy(pre_read, data.data, data.length);
```

Comment:
Could you let me know why we need to copy and to use copied one instead of direct use of data.data?

Reply:
> This is also why the application designed by Realtek uses memcpy to copy the data from the returned buffer to their own buffer and then prints it.
Additionally, I don't think this is good practice considering permanent memory allocations and freeing problems.

Oh here is the answer. So, you want to keep Realtek design. Correct?